lib/db.js

```diff
@@ -909,6 +909,24 @@ module.exports = (
       )
   }
 
+  DB.prototype.createSigninCode = function (uid) {
+    log.trace({ op: 'DB.createSigninCode' })
+
+    return random(config.signinCodeSize)
+      .then(code => {
+        const data = { uid, createdAt: Date.now() }
+        return this.pool.put(`/signinCodes/${code.toString('hex')}`, data)
+          .then(() => code, err => {
+            if (isRecordAlreadyExistsError(err)) {
+              log.warn({ op: 'DB.createSigninCode.duplicate' })
+              return this.createSigninCode(uid)
```

Comment:
Fwiw, there is a similar seems-inifinite-but-isn't-really-infinite recursion of this type already, which we haven't had a problem with:

https://github.com/mozilla/fxa-auth-server/blob/ad706a72499015b650061adfffab7dad02ce3d50/lib/db.js#L531

(full disclosure, I'm responsible for that one too)